url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/table/TableBase.java

```diff
@@ -0,0 +1,272 @@
+package ru.spbau.tinydb.table;
+
+import ru.spbau.tinydb.bufferManager.BufferManager;
+import ru.spbau.tinydb.bufferManager.BufferView;
+import ru.spbau.tinydb.common.DBException;
+
+import java.util.Iterator;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Created by Sergey on 27.10.2014.
+ */
+public class TableBase implements Iterable<ViewWithId> {
+    private final static int POINTERS_SIZE = 8;
+    private final BufferManager bufferManager;
+    private final int firstPage;
+    private final int recordsPerPage;
+    private final int recordSize;
+    private final int bitSetSize;
+    // Header {
+    private int firstFullPage;
+    private int firstNotFullPage;
+    private int rowsCount;
+    // } Header
+
+    public TableBase(BufferManager bufferManager, int firstPage, int recordSize) {
+        this.bufferManager = bufferManager;
+        this.firstPage = firstPage;
+        this.recordSize = recordSize;
+        firstFullPage = 0;
+        firstNotFullPage = 0;
+        rowsCount = 0;
+        // (Available bits count) / (bits per record)
+        // bits per record is 8 * recordSize plus 2 bits for 
+        // 2 bitSets(availability and delete flag)
+        recordsPerPage = (int)(8 * (BufferManager.PAGE_SIZE - POINTERS_SIZE)) / (8 * recordSize + 2);
+        bitSetSize = (recordsPerPage + recordsPerPage + 7) / 8;
+        loadHeader();
+    }
+
+
+    private void writeHeader() {
+        try (BufferView firstPageView = bufferManager.getPage(firstPage)) {
+            firstPageView.setInt(0, firstNotFullPage);
+            firstPageView.setInt(4, firstFullPage);
+            firstPageView.setInt(8, rowsCount);
+            firstPageView.setChanged();
+        }
+
+    }
+
+    private void loadHeader() {
+        try (BufferView firstPageView = bufferManager.getPage(firstPage)) {
+            firstNotFullPage = firstPageView.getInt(0);
+            firstFullPage = firstPageView.getInt(4);
+            rowsCount = firstPageView.getInt(8);
+        }
+    }
+
+    public int insert(byte[] row) throws ExecutionException {
+        if(row.length != recordSize) {
+            throw new IllegalArgumentException();
+        }
+        BufferView pageView;
+        if(firstNotFullPage == 0) {
+            firstNotFullPage = bufferManager.getFreePage();
+            writeHeader();
+            pageView = bufferManager.getPage(firstNotFullPage);
+            pageView.setChanged();
+        } else {
+            pageView = bufferManager.getPage(firstNotFullPage);
+        }
+        int pageIndex = firstNotFullPage;
+
+
+        // writing row
+        BufferView bitSetView = pageView.getSubView(POINTERS_SIZE, bitSetSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *bitSetView*. The resource is referenced at lines 69, 77, 78, 81, 86, 87, 92, 100, 101, 117, 118, 130. The resource is closed at lines 88, 97, 131. There are other execution paths that do not contain closure statements, for example, when an if condition at line 82 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).